src/main/java/org/commcare/cases/util/StorageBackedTreeRoot.java

```diff
@@ -101,40 +100,68 @@ private void collectProcessablePredicates(Vector<XPathExpression> predicates,
+                    for (Enumeration en = indices.keys(); en.hasMoreElements(); ) {
+                        XPathPathExpr expr = (XPathPathExpr)en.nextElement();
+                        if (expr.matches(lookupArg)) {
+                            String filterIndex = translateFilterExpr(expr, (XPathPathExpr)lookupArg, indices);
+
+                            //TODO: We need a way to determine that this value does not also depend on anything in the current context, not
+                            //sure the best way to do that....? Maybe tell the evaluation context to skip out here if it detects a request
+                            //to resolve in a certain area?
+                            Object o = FunctionUtils.unpack(((XPathSelectedFunc)xpe).args[0].eval(evalContext));
+
+                            optimizations.addElement(new IndexedSetMemberLookup(filterIndex, o));
 
                             continue predicate;
                         }
                     }
                 }
             }
+
+
             //There's only one case where we want to keep moving along, and we would have triggered it if it were going to happen,
             //so otherwise, just get outta here.
             break;
         }
     }
 
     private Vector<Integer> processPredicates(Vector<Integer> toRemove,
-                                              Vector<String> keysToFetch,
-                                              Vector<Object> valuesToFetch) {
+                                              Vector<PredicateProfile> profiles) {
         Vector<Integer> selectedElements = null;
         IStorageUtilityIndexed<?> storage = getStorage();
         int predicatesProcessed = 0;
-        while (keysToFetch.size() > 0) {
-            //Get the first set of values.
-            String key = keysToFetch.elementAt(0);
-            Object o = valuesToFetch.elementAt(0);
-            int startCount = keysToFetch.size();
-
-            //Some storage roots will collect common iterative mappings ahead of time,
-            //go check whether this key is loaded into cached memory.
-            Hashtable<String, Integer> keyMapping = getKeyMapping(key);
+        while (profiles.size() > 0) {
+
+            int startCount = profiles.size();
+
+            Hashtable<String, Integer> keyMapping = null;
+
+            if(profiles.elementAt(0) instanceof IndexedValueLookup) {
+
+                IndexedValueLookup valueLookup =
+                        (IndexedValueLookup)profiles.elementAt(0);
+                //Get the first set of values.
+                String key = valueLookup.key;
+
+                //Some storage roots will collect common iterative mappings ahead of time,
+                //go check whether this key is loaded into cached memory.
+                keyMapping = getKeyMapping(key);
+            }
+
+            //TODO: Move key mapping and "Index Match" to be their own optimization implementations
             if (keyMapping != null) {
```

Comment:
The logic is kind of awkward because you need to be able to fall over to the second body _either_ because the keyMapping doesn't exist for this key, or because it's the wrong kind of optimization (IE: If the keyMapping does exist, we want to never execute the second body, but it could not exist for one of two reasons). 

Definitely needs refactoring, but I think I'm going to refactor the optimization set entirely to make this make more sense next (outside of the 2.33 release)